### DISCUSSION TITLE
guile: Question about MACOSX_DEPLOYMENT_TARGET of Homebrew packages.

### DISCUSSION THREAD

**Discussion by toothbrush** [Homebrew/discussions#2114] [Everyday usage] [answered]
Hey folks, first off i apologise if i'm doing something obviously wrong. I've spent a number of days battling to get to a stage where i'm even able to statically link against `libguile` from the [Guile Homebrew package](https://github.com/Homebrew/homebrew-core/blob/master/Formula/guile.rb), in my Xcode project, but i can't seem to squash this final bug.  I have a minimal working example over here: https://github.com/toothbrush/cli-guile. That project links statically against `libguile` and importantly has a setting `MACOSX_DEPLOYMENT_TARGET = 10.15;` because i want users of macOS 10.15 to be able to use my app.

## Update: i understand the issue better, and am looking for suggestions to work around it

Basically, i'm on a Big Sur machine and i'm writing software (in Xcode, for macOS) that i'd like to distribute with `MACOSX_DEPLOYMENT_TARGET = 10.15` so that folks on older macOS are able to download and use my binaries.  However, if i rely on dependencies installed by Homebrew, they're likely to be built for _at least_ my current macOS version.

**Question:** Is there a way to tell Homebrew to use 10.15 packages, even though i'm on newer macOS?

### High-level issue

On one of my macOS computers, compiling the project gives reams (hundreds) of errors like:

```
ld: warning: object file (/usr/local/lib/libguile-3.0.a(libguile_3.0_la-load.o)) was built for newer macOS version (11.0) than being linked (10.15)
```

On another macOS machine (nothing fancy, just `brew install guile` version 3.0.7) those errors aren't present.

### Digging deeper

I worked out that i can use `otool` to get more information about the `.a` files.  I see that there is a difference between the `.a` files on the two computers!  This surprises me since i should think they're bitwise identical if they're "poured" from a Homebrew "bottle".

```shellsession
# On the machine giving the warnings:
$ otool -lv /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a|grep minos|uniq   
    minos 11.0
$ sha256sum  /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a
7b903f24720dc3843e76829d103f03cd66cb914abbb6898f5a3d4e752a5d9e2e  /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a
$ brew info guile
guile: stable 3.0.7 (bottled), HEAD
GNU Ubiquitous Intelligent Language for Extensions
https://www.gnu.org/software/guile/
/usr/local/Cellar/guile/3.0.7 (838 files, 56.8MB) *
  Poured from bottle on 2021-09-08 at 20:59:05
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/guile.rb
License: LGPL-3.0-or-later
==> Dependencies
Build: gnu-sed ✔
Required: bdw-gc ✔, gmp ✔, libffi ✔, libtool ✔, libunistring ✔, pkg-config ✔, readline ✔
==> Options
[snip]
```

```shellsession
# On the computer where stuff works fine:
$ otool -lv /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a|grep minos|uniq
    minos 10.15
$ sha256sum /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a
c3f509d3a5995c328cbffada432d6aec524ba6315f27768396aca5f6306bfd6c  /usr/local/Cellar/guile/3.0.7/lib/libguile-3.0.a
$ brew info guile
guile: stable 3.0.7 (bottled), HEAD
GNU Ubiquitous Intelligent Language for Extensions
https://www.gnu.org/software/guile/
/usr/local/Cellar/guile/3.0.7 (838 files, 56.8MB) *
  Poured from bottle on 2021-05-28 at 10:53:40
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/guile.rb
License: LGPL-3.0-or-later
==> Dependencies
Build: gnu-sed ✔
Required: bdw-gc ✔, gmp ✔, libffi ✔, libtool ✔, libunistring ✔, pkg-config ✔, readline ✔
==> Options, Caveats
[snip]
```

#### Even more details

I'm really surprised that the SHA256 of the library files is different, shouldn't they be the exact same thing extracted from the tarball that Homebrew downloads from gnu.org?  Anyway i'll post the output of `brew config` too, but more generally my question is, what influences the libraries installed by Homebrew? What could be modifying it?

```shellsession
# The computer throwing the warnings:
$ brew config
HOMEBREW_VERSION: 3.2.11-28-g5c60200
ORIGIN: https://github.com/Homebrew/brew
HEAD: 5c602003f29a955146fb9e49df97d5390c6fa804
Last commit: 22 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 28500e1a13cf02b5edfbf06e1ba58a5105990c97
Core tap last commit: 14 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: ["--require-sha"]
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_GITHUB_API: set
HOMEBREW_NO_INSECURE_REDIRECT: set
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: 12.0.5 build 1205
Git: 2.33.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.5.2-x86_64
CLT: 12.5.1.0.1.1623191612
Xcode: 12.5
```

```shellsession
# The computer working the way i want:
$ brew config
HOMEBREW_VERSION: 3.2.11-24-gfd02cfe
ORIGIN: https://github.com/Homebrew/brew
HEAD: fd02cfe65cbceb0c7d7bfce2cb26d38f20840fdc
Last commit: 24 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: ba95d084b25bdcd1460c26ee91b96ca015fa3701
Core tap last commit: 25 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: ["--require-sha"]
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_GITHUB_API: set
HOMEBREW_NO_INSECURE_REDIRECT: set
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 12.0.5 build 1205
Git: 2.32.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.5.2-x86_64
CLT: 12.5.1.0.1.1623191612
Xcode: 12.5.1
```

**Comment by toothbrush** [comment, marked as answer]
**See answer from @carlocab in thread**

Oh, i think i might have a lead – i think on one of the machines i must have installed the `guile` package before upgrading to Big Sur:

```shellsession
# warnings machine
10:24:53 $ ll ~/Library/Caches/Homebrew/downloads/*guile*
-rw-r--r-- 1 me staff 8.1K Sep  8 18:35 /Users/me/Library/Caches/Homebrew/downloads/0b1b3d7479fd3f50617bcce4eb37a5e6b80c1697bb63805fada6773dafb033d7--guile-3.0.7.bottle_manifest.json
-rw-r--r-- 1 me staff  12M May 11 06:05 /Users/me/Library/Caches/Homebrew/downloads/d7a137bc9ec550787f0958bc9933384aec7602a6ff693b2f0ce12a47829c0e58--guile--3.0.7.big_sur.bottle.tar.gz
```

```shellsession
# okay machine
10:25:13 $ ll ~/Library/Caches/Homebrew/downloads/*guile*
-rw-r--r-- 1 me staff 6.3K May 28 10:53 /Users/me/Library/Caches/Homebrew/downloads/0b1b3d7479fd3f50617bcce4eb37a5e6b80c1697bb63805fada6773dafb033d7--guile-3.0.7.bottle_manifest.json
-rw-r--r-- 1 me staff 6.3K May  6 15:08 /Users/me/Library/Caches/Homebrew/downloads/2d49ec2852d40d4d731fcdde231ed904f99aca754aedef04aefb8cd4e0c5c44c--guile-3.0.6.bottle_manifest.json
-rw-r--r-- 1 me staff  12M May 11 06:05 /Users/me/Library/Caches/Homebrew/downloads/927d2790fab48c9bd4dfbe020b30e94987df1e8f54ab60ac55bf84a012da66d4--guile--3.0.7.catalina.bottle.tar.gz
```

**Reply by carlocab** [reply]
This is probably why. You have two different bottles on those two machines.

If you want to link against a `libguile` built on Catalina, you can download the bottle with
```
brew fetch --bottle-tag=catalina guile
```

This will place the bottle in `$(brew --cache)`, though `brew` will tell you where it put the bottle.

I don't recommend this for dynamic linking. At least, if you do use it for that, expect to need a lot of `install_name_tool`.

I'll also mention that Homebrew isn't really designed to build things that are meant to be redistributed elsewhere, so you may run into weird problems. (Though in general static linking is probably fine.)

**Reply by toothbrush** [reply]
Amazing, i'll try that out – thank you so much!

Yeah i'm creating a static binary with a bunch of `.a` things compiled in, only relying on system dylibs.

**Reply by toothbrush** [reply]
Okay that seems to have removed some warnings, excellent!  For posterity, here's what i did:

```shellsession
$ brew fetch --bottle-tag=catalina guile
Downloaded to: /Users/me/Library/Caches/Homebrew/downloads/927d2790fab48c9bd4dfbe020b30e94987df1e8f54ab60ac55bf84a012da66d4--guile--3.0.7.catalina.bottle.tar.gz
SHA256: ee1867daea429b0e7867a30890e07f3c7e4a69d6d483c728c912aea34aa4f83d
$ brew reinstall --force /Users/me/Library/Caches/Homebrew/downloads/927d2790fab48c9bd4dfbe020b30e94987df1e8f54ab60ac55bf84a012da66d4--guile--3.0.7.catalina.bottle.tar.gz
```
